File: library/src/main/java/com/afollestad/iconrequest/LoadResult.java
package com.afollestad.iconrequest;

import android.os.Parcelable;
import android.support.annotation.Nullable;

import com.google.auto.value.AutoValue;

import java.util.ArrayList;
import java.util.List;

@AutoValue
public abstract class LoadResult implements Parcelable {

  static LoadResult create(List<AppModel> apps) {
    return new AutoValue_LoadResult(apps, null);
  }

  static LoadResult create(Exception error) {
    return new AutoValue_LoadResult(new ArrayList<>(0), error);
  }

  public boolean success() {
    return error() != null;
  }

  public abstract List<AppModel> apps();

  @Nullable
  public abstract Exception error();
}

File: library/src/main/java/com/afollestad/iconrequest/LoadResult.kt
package com.afollestad.iconrequest

/** @author Aidan Follestad (afollestad) */
data class LoadResult(
  val apps: List<AppModel> = listOf(),
  val error: Throwable? = null
) {
  val success: Boolean
    get() = error == null
}